{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Using NEAT to evolve solutions to XOR\n",
    "---\n",
    "\n",
    "To use NEAT you must import the following."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from neat3 import config, population, chromosome, genome, visualize\n",
    "from neat3.nn import nn_pure as nn\n",
    "import pickle"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Configuring NEAT\n",
    "\n",
    "In order to run NEAT, you'll need to configure a number of paramters.  There is a configuration file, provided below, where these parameters are set.  \n",
    "\n",
    "The **[phenotype]** section defines the type of neural networks that will be evolved.  You need to ensure that the number of input nodes and output nodes match the requirements of your problem domain.  Typically you will start with zero hidden nodes, and let NEAT add them if necessary to achieve higher fitness. The **nn_activation** defines what activation function will be used.  The function **exp** is a sigmoid and will only return values in the range **[0,1]**.  The function **tanh** will return values in the range **[-1,1]**. Typically for robot control problems we will use **tanh**.\n",
    "\n",
    "The **[genetic]** section defines the size of the population and the maximum fitness threshold.  Both of these should be tuned for your specific problem domain. When deciding on a **pop_size**, keep in mind that the larger the population, the longer the evolution process will take.  However, larger populations tend to yield better results.  You'll need to decide the best tradeoff of size vs speed. The **max_fitness_threshold** is used by NEAT to decide whether to stop evolution early.  If this threshold is ever achieved, NEAT will suspend evolution, otherwise it will continue until the specified number of generations is reached.\n",
    "\n",
    "The other sections tune how NEAT operates.  I tend to leave these at the default settings, but feel free to experiment with them."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting configXOR\n"
     ]
    }
   ],
   "source": [
    "%%file configXOR\n",
    "\n",
    "[phenotype]\n",
    "input_nodes         = 2\n",
    "output_nodes        = 1\n",
    "max_weight          = 30\n",
    "min_weight          = -30\n",
    "feedforward         = 1\n",
    "nn_activation       = exp\n",
    "hidden_nodes        = 0\n",
    "weight_stdev        = 0.9\n",
    "\n",
    "[genetic]\n",
    "pop_size              = 50\n",
    "max_fitness_threshold = 1\n",
    "\n",
    "# Human reasoning\n",
    "prob_addconn          = 0.1\n",
    "prob_addnode          = 0.05\n",
    "prob_mutatebias       = 0.2\n",
    "bias_mutation_power   = 0.5\n",
    "prob_mutate_weight    = 0.9\n",
    "weight_mutation_power = 1.5\n",
    "prob_togglelink       = 0.01\n",
    "elitism               = 1\n",
    "\n",
    "[genotype compatibility]\n",
    "compatibility_threshold = 3.0\n",
    "compatibility_change    = 0.0\n",
    "excess_coeficient       = 1.0\n",
    "disjoint_coeficient     = 1.0\n",
    "weight_coeficient       = 0.4\n",
    "\n",
    "[species]\n",
    "species_size        = 10\n",
    "survival_threshold  = 0.2\n",
    "old_threshold       = 30\n",
    "youth_threshold     = 10\n",
    "old_penalty         = 0.2\n",
    "youth_boost         = 1.2\n",
    "max_stagnation      = 15"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Define a fitness function\n",
    "\n",
    "Create a fitness function that takes a chromosome as input and as a side-effect sets the chromosome's fitness. It does not return anything. A fitness function should be defined such that higher values equate with better fitness.  \n",
    "\n",
    "For this task, we want to define fitness to match how well the network solves the XOR problem. We first calculate the sum-squared error of a network on the XOR task.  Good networks will have low error, so we will define fitness as **1 - sqrt(avgError)**.  In order to evaluate a NEAT chromosome, you must first convert the chromosome into a phenotype (a network).  Then you can activate the network and test its output values."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import math\n",
    "INPUTS = [[0, 0], [0, 1], [1, 0], [1, 1]]\n",
    "OUTPUTS = [0, 1, 1, 0]\n",
    "\n",
    "def eval_individual(chromo):\n",
    "    net = nn.create_ffphenotype(chromo)\n",
    "    error = 0.0\n",
    "    for i, inputs in enumerate(INPUTS):\n",
    "        net.flush() \n",
    "        output = net.sactivate(inputs)\n",
    "        error += (output[0] - OUTPUTS[i])**2\n",
    "    chromo.fitness = 1 - math.sqrt(error/len(OUTPUTS))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Define a population evaluation function\n",
    "\n",
    "NEAT requires a population evaluation function that takes a population as input. A population is defined to be a list of chromosomes.  This function will be called once for every generation of the evolution, and it will evaluate the fitness of every member of the population."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def eval_population(population):\n",
    "    for chromo in population:\n",
    "        eval_individual(chromo)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Evolve\n",
    "\n",
    "The evolve function takes the number of generations to run. Each generation, NEAT will print out statistics about the population, including the current best fitness and the average fitness. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def evolve(n):\n",
    "    config.load(\"configXOR\")\n",
    "    chromosome.node_gene_type = genome.NodeGene\n",
    "\n",
    "    # Tell NEAT that we want to use the above function to evaluate fitness\n",
    "    population.Population.evaluate = eval_population\n",
    "\n",
    "    # Create a population (the size is defined in the configuration file)\n",
    "    pop = population.Population()\n",
    "\n",
    "    # Run NEAT's genetic algorithm for at most 30 epochs\n",
    "    # It will stop if fitness surpasses the max_fitness_threshold in config file\n",
    "    pop.epoch(n, report=True, save_best=True, name=\"XOR\")\n",
    "\n",
    "    # Plots the evolution of the best/average fitness\n",
    "    visualize.plot_stats(pop.stats)\n",
    "\n",
    "    # Visualizes speciation\n",
    "    visualize.plot_species(pop.species_log)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      " ****** Running generation 0 ****** \n",
      "\n",
      "Population's average fitness: 0.39660 stdev: 0.06082\n",
      "Best fitness: 0.4998555915 - size: (0, 2) - species 1 - id 30\n",
      "Species length: 1 totalizing 50 individuals\n",
      "Species ID       : [1]\n",
      "Each species size: [50]\n",
      "Amount to spawn  : [50]\n",
      "Species age      : [0]\n",
      "Species no improv: [0]\n",
      "\n",
      " ****** Running generation 1 ****** \n",
      "\n",
      "Population's average fitness: 0.40501 stdev: 0.06442\n",
      "Best fitness: 0.4998555915 - size: (0, 2) - species 1 - id 30\n",
      "Species length: 1 totalizing 50 individuals\n",
      "Species ID       : [1]\n",
      "Each species size: [50]\n",
      "Amount to spawn  : [50]\n",
      "Species age      : [1]\n",
      "Species no improv: [0]\n",
      "\n",
      " ****** Running generation 2 ****** \n",
      "\n",
      "Population's average fitness: 0.39970 stdev: 0.07279\n",
      "Best fitness: 0.4998555915 - size: (0, 2) - species 1 - id 30\n",
      "Species length: 3 totalizing 50 individuals\n",
      "Species ID       : [1, 2, 3]\n",
      "Each species size: [47, 2, 1]\n",
      "Amount to spawn  : [17, 14, 20]\n",
      "Species age      : [2, 0, 0]\n",
      "Species no improv: [1, 0, 0]\n",
      "Removing 1 excess indiv(s) from the new population\n",
      "\n",
      " ****** Running generation 3 ****** \n",
      "\n",
      "Population's average fitness: 0.38665 stdev: 0.07965\n",
      "Best fitness: 0.4999850268 - size: (1, 4) - species 1 - id 183\n",
      "Species length: 2 totalizing 50 individuals\n",
      "Species ID       : [1, 2]\n",
      "Each species size: [41, 9]\n",
      "Amount to spawn  : [26, 24]\n",
      "Species age      : [3, 1]\n",
      "Species no improv: [2, 0]\n",
      "\n",
      " ****** Running generation 4 ****** \n",
      "\n",
      "Population's average fitness: 0.37976 stdev: 0.08801\n",
      "Best fitness: 0.4999850268 - size: (1, 4) - species 1 - id 183\n",
      "Species length: 2 totalizing 50 individuals\n",
      "Species ID       : [1, 2]\n",
      "Each species size: [26, 24]\n",
      "Amount to spawn  : [26, 24]\n",
      "Species age      : [4, 2]\n",
      "Species no improv: [3, 1]\n",
      "\n",
      " ****** Running generation 5 ****** \n",
      "\n",
      "Population's average fitness: 0.39275 stdev: 0.09239\n",
      "Best fitness: 0.4999964553 - size: (1, 4) - species 1 - id 259\n",
      "Species length: 2 totalizing 50 individuals\n",
      "Species ID       : [1, 2]\n",
      "Each species size: [25, 25]\n",
      "Amount to spawn  : [27, 23]\n",
      "Species age      : [5, 3]\n",
      "Species no improv: [0, 0]\n",
      "\n",
      " ****** Running generation 6 ****** \n",
      "\n",
      "Population's average fitness: 0.43390 stdev: 0.09431\n",
      "Best fitness: 0.5179961340 - size: (1, 5) - species 1 - id 320\n",
      "Species length: 3 totalizing 50 individuals\n",
      "Species ID       : [1, 2, 4]\n",
      "Each species size: [47, 2, 1]\n",
      "Amount to spawn  : [18, 19, 13]\n",
      "Species age      : [6, 4, 0]\n",
      "Species no improv: [0, 0, 0]\n",
      "\n",
      " ****** Running generation 7 ****** \n",
      "\n",
      "Population's average fitness: 0.42348 stdev: 0.09208\n",
      "Best fitness: 0.5179961340 - size: (1, 5) - species 1 - id 320\n",
      "Species length: 3 totalizing 50 individuals\n",
      "Species ID       : [1, 4, 5]\n",
      "Each species size: [35, 11, 4]\n",
      "Amount to spawn  : [18, 12, 19]\n",
      "Species age      : [7, 1, 0]\n",
      "Species no improv: [0, 0, 0]\n",
      "Selecting 1 more indiv(s) to fill up the new population\n",
      "\n",
      " ****** Running generation 8 ****** \n",
      "\n",
      "Population's average fitness: 0.44845 stdev: 0.07554\n",
      "Best fitness: 0.5179961340 - size: (1, 5) - species 1 - id 320\n",
      "Species length: 2 totalizing 50 individuals\n",
      "Species ID       : [1, 4]\n",
      "Each species size: [38, 12]\n",
      "Amount to spawn  : [26, 24]\n",
      "Species age      : [8, 2]\n",
      "Species no improv: [0, 0]\n",
      "\n",
      " ****** Running generation 9 ****** \n",
      "\n",
      "Population's average fitness: 0.45446 stdev: 0.07062\n",
      "Best fitness: 0.5943586243 - size: (1, 5) - species 1 - id 437\n",
      "Species length: 2 totalizing 50 individuals\n",
      "Species ID       : [1, 4]\n",
      "Each species size: [26, 24]\n",
      "Amount to spawn  : [25, 25]\n",
      "Species age      : [9, 3]\n",
      "Species no improv: [0, 0]\n",
      "\n",
      " ****** Running generation 10 ****** \n",
      "\n",
      "Population's average fitness: 0.43392 stdev: 0.08821\n",
      "Best fitness: 0.5943586243 - size: (1, 5) - species 1 - id 437\n",
      "Species length: 3 totalizing 50 individuals\n",
      "Species ID       : [1, 4, 6]\n",
      "Each species size: [24, 25, 1]\n",
      "Amount to spawn  : [13, 17, 19]\n",
      "Species age      : [10, 4, 0]\n",
      "Species no improv: [1, 0, 0]\n",
      "Selecting 1 more indiv(s) to fill up the new population\n",
      "\n",
      " ****** Running generation 11 ****** \n",
      "\n",
      "Population's average fitness: 0.44828 stdev: 0.09289\n",
      "Best fitness: 0.5943586243 - size: (1, 5) - species 1 - id 437\n",
      "Species length: 4 totalizing 50 individuals\n",
      "Species ID       : [1, 4, 6, 7]\n",
      "Each species size: [13, 18, 18, 1]\n",
      "Amount to spawn  : [12, 14, 14, 9]\n",
      "Species age      : [11, 5, 1, 0]\n",
      "Species no improv: [2, 0, 1, 0]\n",
      "Selecting 1 more indiv(s) to fill up the new population\n",
      "\n",
      " ****** Running generation 12 ****** \n",
      "\n",
      "Population's average fitness: 0.45460 stdev: 0.08749\n",
      "Best fitness: 0.6254678735 - size: (1, 4) - species 1 - id 579\n",
      "Species length: 4 totalizing 50 individuals\n",
      "Species ID       : [1, 4, 6, 7]\n",
      "Each species size: [12, 28, 1, 9]\n",
      "Amount to spawn  : [11, 14, 14, 11]\n",
      "Species age      : [12, 6, 2, 1]\n",
      "Species no improv: [3, 0, 2, 0]\n",
      "\n",
      " ****** Running generation 13 ****** \n",
      "\n",
      "Population's average fitness: 0.46742 stdev: 0.08110\n",
      "Best fitness: 0.6254678735 - size: (1, 4) - species 1 - id 579\n",
      "Species length: 4 totalizing 50 individuals\n",
      "Species ID       : [1, 4, 6, 7]\n",
      "Each species size: [11, 13, 14, 12]\n",
      "Amount to spawn  : [11, 13, 13, 13]\n",
      "Species age      : [13, 7, 3, 2]\n",
      "Species no improv: [0, 1, 3, 0]\n",
      "\n",
      " ****** Running generation 14 ****** \n",
      "\n",
      "Population's average fitness: 0.45172 stdev: 0.08333\n",
      "Best fitness: 0.6254678735 - size: (1, 4) - species 1 - id 625\n",
      "Species length: 6 totalizing 50 individuals\n",
      "Species ID       : [1, 4, 6, 7, 8, 9]\n",
      "Each species size: [11, 13, 12, 11, 1, 2]\n",
      "Amount to spawn  : [7, 9, 8, 8, 9, 9]\n",
      "Species age      : [14, 8, 4, 3, 0, 0]\n",
      "Species no improv: [1, 0, 4, 1, 0, 0]\n",
      "\n",
      " ****** Running generation 15 ****** \n",
      "\n",
      "Population's average fitness: 0.48658 stdev: 0.07023\n",
      "Best fitness: 0.6254678735 - size: (1, 4) - species 1 - id 625\n",
      "Species length: 7 totalizing 50 individuals\n",
      "Species ID       : [1, 4, 6, 7, 8, 9, 10]\n",
      "Each species size: [7, 12, 7, 9, 6, 8, 1]\n",
      "Amount to spawn  : [7, 7, 7, 7, 7, 7, 8]\n",
      "Species age      : [15, 9, 5, 4, 1, 1, 0]\n",
      "Species no improv: [0, 0, 5, 2, 1, 1, 0]\n",
      "\n",
      " ****** Running generation 16 ****** \n",
      "\n",
      "Population's average fitness: 0.47677 stdev: 0.08117\n",
      "Best fitness: 0.6254678735 - size: (1, 4) - species 1 - id 720\n",
      "Species length: 7 totalizing 50 individuals\n",
      "Species ID       : [1, 4, 6, 7, 8, 9, 10]\n",
      "Each species size: [7, 12, 7, 7, 2, 7, 8]\n",
      "Amount to spawn  : [7, 6, 7, 7, 8, 7, 7]\n",
      "Species age      : [16, 10, 6, 5, 2, 2, 1]\n",
      "Species no improv: [0, 1, 6, 3, 2, 2, 1]\n",
      "Selecting 1 more indiv(s) to fill up the new population\n",
      "\n",
      " ****** Running generation 17 ****** \n",
      "\n",
      "Population's average fitness: 0.47432 stdev: 0.06924\n",
      "Best fitness: 0.6254678735 - size: (1, 4) - species 1 - id 765\n",
      "Species length: 6 totalizing 50 individuals\n",
      "Species ID       : [1, 4, 6, 7, 9, 10]\n",
      "Each species size: [7, 15, 7, 10, 4, 7]\n",
      "Amount to spawn  : [7, 8, 8, 9, 9, 8]\n",
      "Species age      : [17, 11, 7, 6, 3, 2]\n",
      "Species no improv: [1, 0, 7, 0, 0, 2]\n",
      "Selecting 1 more indiv(s) to fill up the new population\n",
      "\n",
      " ****** Running generation 18 ****** \n",
      "\n",
      "Population's average fitness: 0.49630 stdev: 0.06692\n",
      "Best fitness: 0.6260981886 - size: (1, 4) - species 1 - id 851\n",
      "Species length: 5 totalizing 50 individuals\n",
      "Species ID       : [1, 4, 6, 7, 10]\n",
      "Each species size: [7, 9, 8, 18, 8]\n",
      "Amount to spawn  : [11, 9, 11, 10, 9]\n",
      "Species age      : [18, 12, 8, 7, 3]\n",
      "Species no improv: [0, 0, 8, 1, 3]\n",
      "\n",
      " ****** Running generation 19 ****** \n",
      "\n",
      "Population's average fitness: 0.47224 stdev: 0.07451\n",
      "Best fitness: 0.6260981886 - size: (1, 4) - species 1 - id 851\n",
      "Species length: 6 totalizing 50 individuals\n",
      "Species ID       : [1, 4, 6, 7, 10, 11]\n",
      "Each species size: [11, 8, 11, 10, 9, 1]\n",
      "Amount to spawn  : [6, 8, 9, 9, 8, 9]\n",
      "Species age      : [19, 13, 9, 8, 4, 0]\n",
      "Species no improv: [1, 1, 9, 0, 4, 0]\n",
      "Selecting 1 more indiv(s) to fill up the new population\n",
      "\n",
      " ****** Running generation 20 ****** \n",
      "\n",
      "Population's average fitness: 0.48621 stdev: 0.05207\n",
      "Best fitness: 0.6260981886 - size: (1, 4) - species 1 - id 892\n",
      "Species length: 7 totalizing 50 individuals\n",
      "Species ID       : [1, 4, 6, 7, 10, 11, 12]\n",
      "Each species size: [6, 8, 9, 9, 8, 9, 1]\n",
      "Amount to spawn  : [6, 7, 7, 8, 8, 8, 5]\n",
      "Species age      : [20, 14, 10, 9, 5, 1, 0]\n",
      "Species no improv: [2, 2, 10, 0, 5, 1, 0]\n",
      "Selecting 1 more indiv(s) to fill up the new population\n",
      "\n",
      " ****** Running generation 21 ****** \n",
      "\n",
      "Population's average fitness: 0.46624 stdev: 0.08291\n",
      "Best fitness: 0.6260981886 - size: (1, 4) - species 1 - id 892\n",
      "Species length: 8 totalizing 50 individuals\n",
      "Species ID       : [1, 4, 6, 7, 10, 12, 13, 14]\n",
      "Each species size: [7, 15, 7, 4, 7, 5, 4, 1]\n",
      "Amount to spawn  : [6, 6, 5, 6, 7, 6, 7, 7]\n",
      "Species age      : [21, 15, 11, 10, 6, 1, 0, 0]\n",
      "Species no improv: [3, 3, 11, 1, 6, 0, 0, 0]\n",
      "\n",
      " ****** Running generation 22 ****** \n",
      "\n",
      "Population's average fitness: 0.45428 stdev: 0.08585\n",
      "Best fitness: 0.6260981886 - size: (1, 4) - species 1 - id 892\n",
      "Species length: 8 totalizing 50 individuals\n",
      "Species ID       : [1, 4, 6, 7, 10, 12, 13, 14]\n",
      "Each species size: [6, 7, 5, 12, 6, 6, 1, 7]\n",
      "Amount to spawn  : [6, 6, 6, 6, 6, 7, 7, 6]\n",
      "Species age      : [22, 16, 12, 11, 7, 2, 1, 1]\n",
      "Species no improv: [4, 4, 0, 2, 7, 0, 0, 1]\n",
      "\n",
      " ****** Running generation 23 ****** \n",
      "\n",
      "Population's average fitness: 0.47589 stdev: 0.07391\n",
      "Best fitness: 0.6260981886 - size: (1, 4) - species 1 - id 892\n",
      "Species length: 7 totalizing 50 individuals\n",
      "Species ID       : [1, 4, 6, 7, 12, 13, 14]\n",
      "Each species size: [6, 14, 6, 12, 7, 1, 4]\n",
      "Amount to spawn  : [6, 6, 7, 7, 8, 8, 7]\n",
      "Species age      : [23, 17, 13, 12, 3, 2, 2]\n",
      "Species no improv: [5, 5, 1, 3, 0, 0, 2]\n",
      "Selecting 1 more indiv(s) to fill up the new population\n",
      "\n",
      " ****** Running generation 24 ****** \n",
      "\n",
      "Population's average fitness: 0.45183 stdev: 0.08799\n",
      "Best fitness: 0.6260981886 - size: (1, 4) - species 1 - id 1064\n",
      "Species length: 9 totalizing 50 individuals\n",
      "Species ID       : [1, 4, 6, 7, 12, 13, 14, 15, 16]\n",
      "Each species size: [6, 7, 6, 7, 7, 8, 7, 1, 1]\n",
      "Amount to spawn  : [4, 5, 6, 5, 6, 6, 6, 4, 7]\n",
      "Species age      : [24, 18, 14, 13, 4, 3, 3, 0, 0]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Species no improv: [6, 6, 0, 4, 1, 1, 3, 0, 0]\n",
      "Selecting 1 more indiv(s) to fill up the new population\n",
      "\n",
      " ****** Running generation 25 ****** \n",
      "\n",
      "Population's average fitness: 0.44886 stdev: 0.09305\n",
      "Best fitness: 0.6260981886 - size: (1, 4) - species 1 - id 1064\n",
      "Species length: 9 totalizing 50 individuals\n",
      "Species ID       : [1, 4, 6, 7, 12, 13, 14, 15, 16]\n",
      "Each species size: [4, 6, 9, 6, 12, 6, 5, 1, 1]\n",
      "Amount to spawn  : [6, 5, 5, 6, 7, 6, 7, 4, 4]\n",
      "Species age      : [25, 19, 15, 14, 5, 4, 4, 1, 1]\n",
      "Species no improv: [7, 7, 1, 5, 2, 2, 4, 1, 1]\n",
      "\n",
      " ****** Running generation 26 ****** \n",
      "\n",
      "Population's average fitness: 0.45017 stdev: 0.10080\n",
      "Best fitness: 0.6260981886 - size: (1, 4) - species 1 - id 1064\n",
      "Species length: 8 totalizing 50 individuals\n",
      "Species ID       : [1, 4, 6, 7, 12, 13, 15, 16]\n",
      "Each species size: [6, 12, 5, 6, 7, 6, 4, 4]\n",
      "Amount to spawn  : [7, 6, 7, 7, 7, 8, 5, 5]\n",
      "Species age      : [26, 20, 16, 15, 6, 5, 2, 2]\n",
      "Species no improv: [8, 8, 2, 6, 3, 3, 0, 2]\n",
      "Removing 2 excess indiv(s) from the new population\n",
      "\n",
      " ****** Running generation 27 ****** \n",
      "\n",
      "Population's average fitness: 0.46011 stdev: 0.08903\n",
      "Best fitness: 0.6260981886 - size: (1, 4) - species 1 - id 1191\n",
      "Species length: 9 totalizing 50 individuals\n",
      "Species ID       : [1, 4, 6, 7, 12, 13, 15, 16, 17]\n",
      "Each species size: [7, 5, 7, 7, 7, 8, 5, 3, 1]\n",
      "Amount to spawn  : [6, 5, 5, 5, 6, 6, 6, 4, 7]\n",
      "Species age      : [27, 21, 17, 16, 7, 6, 3, 3, 0]\n",
      "Species no improv: [9, 9, 3, 7, 4, 4, 0, 3, 0]\n",
      "\n",
      " ****** Running generation 28 ****** \n",
      "\n",
      "Population's average fitness: 0.48232 stdev: 0.08515\n",
      "Best fitness: 0.6260981886 - size: (1, 4) - species 1 - id 1191\n",
      "Species length: 9 totalizing 50 individuals\n",
      "Species ID       : [1, 4, 6, 7, 12, 13, 15, 16, 17]\n",
      "Each species size: [6, 6, 5, 5, 6, 6, 6, 4, 6]\n",
      "Amount to spawn  : [6, 5, 5, 5, 6, 6, 6, 4, 6]\n",
      "Species age      : [28, 22, 18, 17, 8, 7, 4, 4, 1]\n",
      "Species no improv: [10, 10, 4, 8, 0, 5, 0, 4, 1]\n",
      "Selecting 1 more indiv(s) to fill up the new population\n",
      "\n",
      " ****** Running generation 29 ****** \n",
      "\n",
      "Population's average fitness: 0.46633 stdev: 0.08473\n",
      "Best fitness: 0.6260981886 - size: (1, 4) - species 1 - id 1191\n",
      "Species length: 10 totalizing 50 individuals\n",
      "Species ID       : [1, 4, 6, 7, 12, 13, 15, 16, 17, 18]\n",
      "Each species size: [7, 5, 5, 5, 6, 6, 5, 4, 6, 1]\n",
      "Amount to spawn  : [4, 5, 5, 5, 5, 6, 5, 4, 5, 6]\n",
      "Species age      : [29, 23, 19, 18, 9, 8, 5, 5, 2, 0]\n",
      "Species no improv: [11, 11, 5, 9, 1, 6, 0, 5, 2, 0]\n"
     ]
    }
   ],
   "source": [
    "evolve(30)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Evaluating the results of evolution\n",
    "\n",
    "NEAT will save several plots, one showing the best and average fitness over time, and another showing how NEAT's population speciated over time. These plots will have **.svg** extensions and can be opened from a terminal window using **eog**. Try viewing these plots.\n",
    "\n",
    "In addition, NEAT will save the best chromosome present in the population at the end of each generation. To test a particular chromosome file, you can use the following function. This will also generate a new **.svg** file that includes a visualization of the network's topology.  Try viewing this plot which will have a filename that begins with **phenotype**."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def eval_best(chromo_file):\n",
    "    fp = open(chromo_file, \"rb\")\n",
    "    chromo = pickle.load(fp)\n",
    "    fp.close()\n",
    "    print(chromo)\n",
    "    visualize.draw_net(chromo, \"_\"+ chromo_file)\n",
    "    \n",
    "    print('\\nNetwork output:')\n",
    "    brain = nn.create_ffphenotype(chromo)\n",
    "    for i, inputs in enumerate(INPUTS):\n",
    "        output = brain.sactivate(inputs)\n",
    "        print(\"%1.5f \\t %1.5f\" %(OUTPUTS[i], output[0]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Nodes:\n",
      "\tNode  1  INPUT, bias  0, response 4.924273\n",
      "\tNode  2  INPUT, bias  0, response 4.924273\n",
      "\tNode  3 OUTPUT, bias 1.56438882, response 4.924273\n",
      "\tNode  4 HIDDEN, bias 0.95296582, response 4.94640357\n",
      "Connections:\n",
      "\tIn  1, Out  3, Weight +0.40704, Enabled, Innov 1\n",
      "\tIn  2, Out  3, Weight +0.67050, Disabled, Innov 2\n",
      "\tIn  2, Out  4, Weight -2.18989, Enabled, Innov 3\n",
      "\tIn  4, Out  3, Weight -1.88029, Enabled, Innov 4\n",
      "\tIn  1, Out  4, Weight +3.32298, Enabled, Innov 5\n",
      "Node order: [4]\n",
      "\n",
      "Network output:\n",
      "0.00000 \t 0.18645\n",
      "1.00000 \t 0.99954\n",
      "1.00000 \t 0.61036\n",
      "0.00000 \t 0.61043\n"
     ]
    }
   ],
   "source": [
    "eval_best(\"XOR_best_chromo_29\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "### Experiments\n",
    "\n",
    "Run 10 experiments on the XOR problem using a popultion of size 50.  For each experiment, record whether NEAT evolved a successful solution, and the number of hidden nodes it generated.  \n",
    "\n",
    "Run 10 more experiments on the XOR problem using a different population size (you can try smaller or larger, whichever you prefer).  Record the same information as before, and compare the results."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Use git to commit, add, and push\n",
    "\n",
    "Save this notebook before moving on."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
